File: geometry.py
"""Shared physical camera model for preview, export, and auto correction."""

from __future__ import annotations

import math

import cv2
import numpy as np


FULL_FRAME_DIAGONAL_MM = math.hypot(36.0, 24.0)
STATE_VERSION = 1
CROP_POLICY = "conservative_same_aspect_v1"


def make_intrinsics(width, height, focal_35mm=None):
    """Return centered square-pixel intrinsics, preferring EXIF 35mm focal length."""
    width = float(width)
    height = float(height)
    if focal_35mm is not None and np.isfinite(focal_35mm) and focal_35mm > 0:
        focal_px = float(focal_35mm) * math.hypot(width, height) / FULL_FRAME_DIAGONAL_MM
        source = "exif_35mm"
    else:
        focal_px = max(width, height)
        source = "fallback_max_dimension"
    return {
        "fx": focal_px,
        "fy": focal_px,
        "cx": width / 2.0,
        "cy": height / 2.0,
        "source": source,
        "focal_35mm": float(focal_35mm) if focal_35mm is not None else None,
    }


def intrinsics_matrix(intrinsics):
    return np.array(
        [
            [float(intrinsics["fx"]), 0.0, float(intrinsics["cx"])],
            [0.0, float(intrinsics["fy"]), float(intrinsics["cy"])],
            [0.0, 0.0, 1.0],
        ],
        dtype=np.float64,
    )


def project_to_rotation(matrix):
    matrix = np.asarray(matrix, dtype=np.float64).reshape(3, 3)
    if not np.all(np.isfinite(matrix)):
        raise ValueError("rotation contains non-finite values")
    u, _singular, vt = np.linalg.svd(matrix)
    rotation = u @ vt
    if np.linalg.det(rotation) < 0:
        u[:, -1] *= -1
        rotation = u @ vt
    return rotation


def rotation_angle_deg(rotation):
    rotation = np.asarray(rotation, dtype=np.float64).reshape(3, 3)
    cosine = np.clip((np.trace(rotation) - 1.0) / 2.0, -1.0, 1.0)
    return float(np.degrees(np.arccos(cosine)))


def rotation_from_homography(homography, intrinsics):
    k = intrinsics_matrix(intrinsics)
    candidate = np.linalg.inv(k) @ np.asarray(homography, dtype=np.float64) @ k
    scale = np.cbrt(abs(float(np.linalg.det(candidate))))
    if not np.isfinite(scale) or scale < 1e-12:
        raise ValueError("homography is singular")
    return project_to_rotation(candidate / scale)


def interpolate_rotation(rotation, strength):
    """Interpolate identity -> rotation on SO(3), never in homography space."""
    strength = float(strength)
    if not np.isfinite(strength):
        raise ValueError("strength must be finite")
    rotation = project_to_rotation(rotation)
    rotvec, _ = cv2.Rodrigues(rotation)
    interpolated, _ = cv2.Rodrigues(rotvec * strength)
    return project_to_rotation(interpolated)


def rotation_from_vector(rotation_vector):
    vector = np.asarray(rotation_vector, dtype=np.float64).reshape(3, 1)
    if not np.all(np.isfinite(vector)):
        raise ValueError("rotation vector contains non-finite values")
    rotation, _ = cv2.Rodrigues(vector)
    return project_to_rotation(rotation)


def compose_rotation(rotation_vector, rotation):
    return project_to_rotation(rotation_from_vector(rotation_vector) @ rotation)


def _apply_homography(matrix, points):
    points = np.asarray(points, dtype=np.float64).reshape(-1, 2)
    homogeneous = np.column_stack([points, np.ones(len(points))])
    projected = (np.asarray(matrix, dtype=np.float64) @ homogeneous.T).T
    denominators = projected[:, 2]
    if np.any(np.abs(denominators) < 1e-9):
        raise ValueError("projection crosses infinity")
    return projected[:, :2] / denominators[:, None], denominators


def _fit_aspect_float(crop, target_aspect):
    x0, y0, x1, y1 = [float(value) for value in crop]
    width = x1 - x0
    height = y1 - y0
    if width <= 0 or height <= 0:
        raise ValueError("crop is empty")
    if width / height > target_aspect:
        new_width = height * target_aspect
        center = (x0 + x1) / 2.0
        x0, x1 = center - new_width / 2.0, center + new_width / 2.0
    else:
        new_height = width / target_aspect
        center = (y0 + y1) / 2.0
        y0, y1 = center - new_height / 2.0, center + new_height / 2.0
    return [x0, y0, x1, y1]


def conservative_crop(transformed_corners, target_aspect):
    tl, tr, br, bl = np.asarray(transformed_corners, dtype=np.float64).reshape(4, 2)
    crop = [
        max(tl[0], bl[0]),
        max(tl[1], tr[1]),
        min(tr[0], br[0]),
        min(br[1], bl[1]),
    ]
    return _fit_aspect_float(crop, float(target_aspect))


def _crop_inside_quad(crop, quad, tolerance=1e-5):
    x0, y0, x1, y1 = crop
    corners = ((x0, y0), (x1, y0), (x1, y1), (x0, y1))
    contour = np.asarray(quad, dtype=np.float32).reshape(-1, 1, 2)
    return all(cv2.pointPolygonTest(contour, point, False) >= -tolerance for point in corners)


def compute_view(rotation, intrinsics, image_size, crop=None):
    """Build a fixed-size warp and a crop containing only source pixels."""
    width, height = [float(value) for value in image_size]
    if width <= 1 or height <= 1:
        raise ValueError("image_size must be positive")
    rotation = project_to_rotation(rotation)
    k = intrinsics_matrix(intrinsics)
    homography = k @ rotation @ np.linalg.inv(k)
    image_corners = np.array(
        [[0.0, 0.0], [width, 0.0], [width, height], [0.0, height]],
        dtype=np.float64,
    )
    transformed, denominators = _apply_homography(homography, image_corners)
    if not (np.all(denominators > 0) or np.all(denominators < 0)):
        raise ValueError("projection flips across the camera plane")
    w_ratio = float(np.min(np.abs(denominators)) / np.max(np.abs(denominators)))

    canonical_crop = conservative_crop(transformed, width / height)
    if crop is None:
        crop = canonical_crop
    else:
        crop = [float(value) for value in crop]
        if len(crop) != 4 or not np.all(np.isfinite(crop)):
            raise ValueError("crop must contain four finite values")
        crop = _fit_aspect_float(crop, width / height)
        if not _crop_inside_quad(crop, transformed):
            raise ValueError("crop extends outside transformed source pixels")

    x0, y0, x1, y1 = crop
    crop_width = x1 - x0
    crop_height = y1 - y0
    sx = width / crop_width
    sy = height / crop_height
    if abs(sx - sy) > 1e-6 * max(1.0, sx, sy):
        raise ValueError("crop would require non-uniform scaling")
    view = np.array(
        [[sx, 0.0, -sx * x0], [0.0, sy, -sy * y0], [0.0, 0.0, 1.0]],
        dtype=np.float64,
    ) @ homography
    return {
        "matrix": view,
        "homography": homography,
        "crop": crop,
        "canonical_crop": canonical_crop,
        "transformed_corners": transformed,
        "theta_deg": rotation_angle_deg(rotation),
        "projective_w_ratio": w_ratio,
    }


def make_correction_state(rotation, intrinsics, image_size, crop=None):
    rotation = project_to_rotation(rotation)
    view = compute_view(rotation, intrinsics, image_size, crop=crop)
    return {
        "version": STATE_VERSION,
        "rotation": rotation.reshape(-1).tolist(),
        "intrinsics": dict(intrinsics),
        "crop": list(view["crop"]),
        "crop_policy": CROP_POLICY,
    }


def validate_correction_state(state, image_size, max_rotation_deg=45.0,
                              min_projective_w_ratio=0.22):
    try:
        if not isinstance(state, dict) or state.get("version") != STATE_VERSION:
            raise ValueError("unsupported correction state version")
        rotation_input = np.asarray(state.get("rotation"), dtype=np.float64).reshape(3, 3)
        if not np.all(np.isfinite(rotation_input)):
            raise ValueError("rotation contains non-finite values")
        orthogonality_error = float(np.linalg.norm(rotation_input.T @ rotation_input - np.eye(3)))
        determinant = float(np.linalg.det(rotation_input))
        if orthogonality_error > 1e-5 or abs(determinant - 1.0) > 1e-5:
            raise ValueError("rotation is not in SO(3)")
        rotation = project_to_rotation(rotation_input)
        theta_deg = rotation_angle_deg(rotation)
        if theta_deg > float(max_rotation_deg):
            raise ValueError("rotation exceeds the safety envelope")
        intrinsics = state.get("intrinsics")
        if not isinstance(intrinsics, dict):
            raise ValueError("intrinsics are missing")
        view = compute_view(rotation, intrinsics, image_size, crop=state.get("crop"))
        if view["projective_w_ratio"] < float(min_projective_w_ratio):
            raise ValueError("projection is too close to degeneracy")
        return {
            "accepted": True,
            "reason": None,
            "theta_deg": theta_deg,
            "orthogonality_error": orthogonality_error,
            "determinant": determinant,
            "projective_w_ratio": view["projective_w_ratio"],
            "view": view,
        }
    except (KeyError, TypeError, ValueError, np.linalg.LinAlgError) as error:
        return {
            "accepted": False,
            "reason": str(error),
            "theta_deg": None,
            "orthogonality_error": None,
            "determinant": None,
            "projective_w_ratio": None,
            "view": None,
        }


def warp_with_state(image, state, max_rotation_deg=45.0):
    height, width = image.shape[:2]
    validation = validate_correction_state(
        state,
        [width, height],
        max_rotation_deg=max_rotation_deg,
    )
    if not validation["accepted"]:
        raise ValueError(validation["reason"])
    matrix = validation["view"]["matrix"]
    return cv2.warpPerspective(image, matrix, (width, height))

File: webapp/geometry.js
(function (root) {
  'use strict';

  const STATE_VERSION = 1;
  const CROP_POLICY = 'conservative_same_aspect_v1';

  function multiply(a, b) {
    const result = new Array(9).fill(0);
    for (let row = 0; row < 3; row++) {
      for (let col = 0; col < 3; col++) {
        for (let k = 0; k < 3; k++) result[row * 3 + col] += a[row * 3 + k] * b[k * 3 + col];
      }
    }
    return result;
  }

  function inverse(m) {
    const adj = [
      m[4] * m[8] - m[5] * m[7], m[2] * m[7] - m[1] * m[8], m[1] * m[5] - m[2] * m[4],
      m[5] * m[6] - m[3] * m[8], m[0] * m[8] - m[2] * m[6], m[2] * m[3] - m[0] * m[5],
      m[3] * m[7] - m[4] * m[6], m[1] * m[6] - m[0] * m[7], m[0] * m[4] - m[1] * m[3],
    ];
    const det = m[0] * adj[0] + m[1] * adj[3] + m[2] * adj[6];
    if (!Number.isFinite(det) || Math.abs(det) < 1e-12) throw new Error('matrix is singular');
    return adj.map(value => value / det);
  }

  function apply(matrix, point) {
    const x = matrix[0] * point[0] + matrix[1] * point[1] + matrix[2];
    const y = matrix[3] * point[0] + matrix[4] * point[1] + matrix[5];
    const w = matrix[6] * point[0] + matrix[7] * point[1] + matrix[8];
    if (!Number.isFinite(w) || Math.abs(w) < 1e-9) throw new Error('projection crosses infinity');
    return [x / w, y / w];
  }

  function rotationAngleDeg(rotation) {
    const cosine = Math.max(-1, Math.min(1, (rotation[0] + rotation[4] + rotation[8] - 1) / 2));
    return Math.acos(cosine) * 180 / Math.PI;
  }

  function rotationFromVector(vector) {
    const theta = Math.hypot(vector[0], vector[1], vector[2]);
    if (theta < 1e-12) return [1, 0, 0, 0, 1, 0, 0, 0, 1];
    const x = vector[0] / theta, y = vector[1] / theta, z = vector[2] / theta;
    const c = Math.cos(theta), s = Math.sin(theta), t = 1 - c;
    return [
      t*x*x + c,   t*x*y - s*z, t*x*z + s*y,
      t*x*y + s*z, t*y*y + c,   t*y*z - s*x,
      t*x*z - s*y, t*y*z + s*x, t*z*z + c,
    ];
  }

  function composeRotation(rotationVector, rotation) {
    return multiply(rotationFromVector(rotationVector), rotation);
  }

  function intrinsicsMatrix(intrinsics) {
    return [
      Number(intrinsics.fx), 0, Number(intrinsics.cx),
      0, Number(intrinsics.fy), Number(intrinsics.cy),
      0, 0, 1,
    ];
  }

  function rotationHomography(rotation, intrinsics) {
    const k = intrinsicsMatrix(intrinsics);
    return multiply(k, multiply(rotation, inverse(k)));
  }

  function cameraRotationForDrag(initialRotation, displacement, intrinsics) {
    const dx = Number(displacement[0]), dy = Number(displacement[1]);
    const fx = Number(intrinsics.fx), fy = Number(intrinsics.fy);
    if (![dx, dy, fx, fy].every(Number.isFinite) || !(fx > 0 && fy > 0)) {
      throw new Error('drag and focal lengths must be finite');
    }
    const pitch = -Math.atan(dy / fy);
    const yaw = Math.atan(dx / fx);
    return composeRotation([pitch, yaw, 0], initialRotation);
  }

  function fitAspect(crop, targetAspect) {
    let [x0, y0, x1, y1] = crop.map(Number);
    const width = x1 - x0, height = y1 - y0;
    if (!(width > 0 && height > 0)) throw new Error('crop is empty');
    if (width / height > targetAspect) {
      const newWidth = height * targetAspect, center = (x0 + x1) / 2;
      x0 = center - newWidth / 2; x1 = center + newWidth / 2;
    } else {
      const newHeight = width / targetAspect, center = (y0 + y1) / 2;
      y0 = center - newHeight / 2; y1 = center + newHeight / 2;
    }
    return [x0, y0, x1, y1];
  }

  function conservativeCrop(corners, targetAspect) {
    const [tl, tr, br, bl] = corners;
    return fitAspect([
      Math.max(tl[0], bl[0]),
      Math.max(tl[1], tr[1]),
      Math.min(tr[0], br[0]),
      Math.min(br[1], bl[1]),
    ], targetAspect);
  }

  function computeView(rotation, intrinsics, imageSize, crop) {
    const [width, height] = imageSize.map(Number);
    const homography = rotationHomography(rotation, intrinsics);
    const sourceCorners = [[0, 0], [width, 0], [width, height], [0, height]];
    const denominators = sourceCorners.map(([x, y]) => homography[6] * x + homography[7] * y + homography[8]);
    const sameSide = denominators.every(value => value > 0) || denominators.every(value => value < 0);
    if (!sameSide) throw new Error('projection flips across the camera plane');
    const absW = denominators.map(Math.abs);
    const projectiveWRatio = Math.min(...absW) / Math.max(...absW);
    const transformedCorners = sourceCorners.map(point => apply(homography, point));
    const canonicalCrop = conservativeCrop(transformedCorners, width / height);
    const selectedCrop = crop ? fitAspect(crop, width / height) : canonicalCrop;
    const [x0, y0, x1, y1] = selectedCrop;
    const sx = width / (x1 - x0), sy = height / (y1 - y0);
    if (Math.abs(sx - sy) > 1e-6 * Math.max(1, sx, sy)) throw new Error('crop requires non-uniform scaling');
    const framing = [sx, 0, -sx*x0, 0, sy, -sy*y0, 0, 0, 1];
    return {
      matrix: multiply(framing, homography),
      homography,
      crop: selectedCrop,
      canonicalCrop,
      transformedCorners,
      thetaDeg: rotationAngleDeg(rotation),
      projectiveWRatio,
    };
  }

  function makeState(rotation, intrinsics, imageSize, crop) {
    const view = computeView(rotation, intrinsics, imageSize, crop);
    return {
      version: STATE_VERSION,
      rotation: rotation.slice(),
      intrinsics: { ...intrinsics },
      crop: view.crop.slice(),
      crop_policy: CROP_POLICY,
    };
  }

  function toCssMatrix3d(matrix) {
    return `matrix3d(${[
      matrix[0], matrix[3], 0, matrix[6],
      matrix[1], matrix[4], 0, matrix[7],
      0, 0, 1, 0,
      matrix[2], matrix[5], 0, matrix[8],
    ].join(',')})`;
  }

  const api = {
    STATE_VERSION,
    CROP_POLICY,
    apply,
    cameraRotationForDrag,
    composeRotation,
    computeView,
    makeState,
    multiply,
    rotationAngleDeg,
    rotationFromVector,
    rotationHomography,
    toCssMatrix3d,
  };
  root.PerspectiveGeometry = api;
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
})(typeof window !== 'undefined' ? window : globalThis);
